src/H5FDonion_history.c

```diff
@@ -0,0 +1,297 @@
+ * Function:    H5FD__onion_write_history
+ *
+ * Purpose:     Encode and write history to file at the given address.
+ *
+ * Returns:     Success:    Number of bytes written to destination file (always non-zero)
+ *              Failure:    0
+ *-----------------------------------------------------------------------------
+ */
+uint64_t
+H5FD__onion_write_history(H5FD_onion_history_t *history, H5FD_t *file, haddr_t off_start,
+                          haddr_t filesize_curr)
+{
+    uint32_t       _sum      = 0; /* Required by the API call but unused here */
+    uint64_t       size      = 0;
+    unsigned char *buf       = NULL;
+    uint64_t       ret_value = 0;
+
+    FUNC_ENTER_PACKAGE;
+
+    if (NULL == (buf = H5MM_malloc(H5FD__ONION_ENCODED_SIZE_HISTORY +
+                                   (H5FD__ONION_ENCODED_SIZE_RECORD_POINTER * history->n_revisions))))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, 0, "can't allocate buffer for updated history")
+
+    if (0 == (size = H5FD__onion_history_encode(history, buf, &_sum)))
+        HGOTO_ERROR(H5E_VFL, H5E_BADVALUE, 0, "problem encoding updated history")
+
+    if ((size + off_start > filesize_curr) && (H5FD_set_eoa(file, H5FD_MEM_DRAW, off_start + size) < 0))
+        HGOTO_ERROR(H5E_VFL, H5E_CANTSET, 0, "can't modify EOA for updated history")
+
+    if (H5FD_write(file, H5FD_MEM_DRAW, off_start, size, buf) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_WRITEERROR, 0, "can't write history as intended")
+
+    ret_value = size;
+
+done:
+    H5MM_xfree(buf);
+
+    FUNC_LEAVE_NOAPI(ret_value);
+} /* end H5FD__onion_write_history() */
+
+/*-----------------------------------------------------------------------------
+ * Function:    H5FD__onion_history_decode
+ *
+ * Purpose:     Attempt to read a buffer and store it as a history
+ *              structure.
+ *
+ *              Implementation must correspond with
+ *              H5FD__onion_history_encode().
+ *
+ *              MUST BE CALLED TWICE:
+ *              On the first call, n_records in the destination structure must
+ *              be zero, and record_locs be NULL.
+ *
+ *              If the buffer is well-formed, the destination structure is
+ *              tentatively populated with fixed-size values, and the number of
+ *              bytes read are returned.
+ *
+ *              Prior to the second call, the user must allocate space for
+ *              record_locs to hold n_records record-pointer structs.
+ *
+ *              Then the decode operation is called a second time, and all
+ *              components will be populated (and again number of bytes read is
+ *              returned).
+ *
+ * Return:      Success:    Number of bytes read from buffer
+ *              Failure:    0
+ *-----------------------------------------------------------------------------
+ */
+size_t
+H5FD__onion_history_decode(unsigned char *buf, H5FD_onion_history_t *history)
```

Comment:
Comment for future reference:

The decode and encode routines in this file should probably be aware of how large buf is so we don't run off the end of it on accident.